File: 0707-Design Linked List/0707-Design Linked List.cpp
class MyLinkedList {
public:
    /** Initialize your data structure here. */
    MyLinkedList() {
        head = nullptr;
        tail = nullptr;
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    int get(int index) {
        Node *p = head;
        for (int i = 0; i < index; ++i) {
            if (p == nullptr) {
                return -1;
            }
            p = p->next;
        }
        return p ? p->val : -1;        
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    void addAtHead(int val) {
        Node* node = new Node(val);
        if (head == nullptr) {
            tail = node;
        }
        node->next = head;
        head = node;
    }
    
    /** Append a node of value val to the last element of the linked list. */
    void addAtTail(int val) {
        Node * node = new Node(val);
        if (tail == nullptr) {
            head = node;
        }
        else {
            tail->next = node;
        }
        tail = node;
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    void addAtIndex(int index, int val) {
        if (index == 0) {
            addAtHead(val);
        }
        else {
            Node *p = head;
            for (int i = 0; i < index - 1; ++i) {
                if (p == nullptr) {
                    return;
                }
                p = p->next;
            }
            if (p == nullptr) {
                return;
            }
            Node *node = new Node(val);
            Node *next = p->next;
            p->next = node;
            node->next = next;
            if (next == nullptr) {
                tail = node;
            }
        }        
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    void deleteAtIndex(int index) {
        if (index == 0) {
            if (head == nullptr) {
                return;
            }
            Node *node = head;
            head = head->next;
            if (head == nullptr) {
                tail = nullptr;
            }
            delete node;
        }
        else {
            Node *p = head;
            for (int i = 0; i < index - 1; ++i) {
                if (p == nullptr) {
                    return;
                }
                p = p->next;
            }
            if (p == nullptr || p->next == nullptr) {
                return;
            }
            Node *next = p->next;
            p->next = next->next;
            if (tail == next) {
                tail = p;
            }
            delete next;
        }
        
    }
    
private:
    struct Node {
        int val;
        Node *next;
        Node(int v): val(v), next(nullptr){}
    };
    Node *head, *tail;
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */

File: 0701-0800/0707-Design Linked List/0707-Design Linked List.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

        
class MyLinkedList:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.head = self.tail = None
        

    def get(self, index: int) -> int:
        """
        Get the value of the index-th node in the linked list. If the index is invalid, return -1.
        """
        if index < 0:
            return -1
        p = self.head
        for _ in range(index):
            if p is None:
                return -1
            p = p.next
        return p.val if p else -1
        

    def addAtHead(self, val: int) -> None:
        """
        Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
        """
        node = ListNode(val)
        if self.tail is None:
            self.tail = node
        node.next = self.head
        self.head = node

    def addAtTail(self, val: int) -> None:
        """
        Append a node of value val to the last element of the linked list.
        """
        node = ListNode(val)
        if self.tail is None:
            self.head = node
        else:
            self.tail.next = node
            
        self.tail = node

    def addAtIndex(self, index: int, val: int) -> None:
        """
        Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
        """
        if index < 0:
            return
        if index == 0:
            self.addAtHead(val)
        else:
            prev = self.head
            for _ in range(index - 1):
                if prev is None:
                    return
                prev = prev.next
            if prev is None:
                return
            node = ListNode(val)
            next = prev.next
            prev.next = node
            if next is None:
                self.tail = node
            else:
                node.next = next


    def deleteAtIndex(self, index: int) -> None:
        """
        Delete the index-th node in the linked list, if the index is valid.
        """
        if index < 0:
            return
        if index == 0:
            if self.head is None:
                return
            self.head = self.head.next
            if self.head is None:
                self.tail = None
        else:
            prev = self.head
            for _ in range(index - 1):
                if prev is None:
                    return
                prev = prev.next
            if prev is None or prev.next is None:
                return
            next = prev.next
            prev.next = next.next
            if self.tail == next:
                self.tail = prev


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
